core/embed/trezorhal/optiga/optiga.c

```diff
@@ -634,6 +733,100 @@ int optiga_pin_verify(OPTIGA_UI_PROGRESS ui_progress,
   return OPTIGA_SUCCESS;
 }
 
+int optiga_pin_verify(OPTIGA_UI_PROGRESS ui_progress,
+                      uint8_t stretched_pin[OPTIGA_PIN_SECRET_SIZE]) {
+  // Stretch the PIN more with stretching secrets from the Optiga.
+  optiga_result res = optiga_pin_stretch_secret(ui_progress, stretched_pin);
+  if (res != OPTIGA_SUCCESS) {
+    return res;
+  }
+
+  // Process the stretched PIN using a one-way function before sending it to the
+  // Optiga.
+  uint8_t digest[OPTIGA_PIN_SECRET_SIZE] = {0};
+  hmac_sha256(stretched_pin, OPTIGA_PIN_SECRET_SIZE, NULL, 0, digest);
+
+  // HMAC the digest with the key in OID_PIN_HMAC.
+  uint8_t hmac_buffer[ENCRYPT_SYM_PREFIX_SIZE + OPTIGA_PIN_SECRET_SIZE] = {0};
+  size_t size = 0;
+  res = optiga_encrypt_sym(OPTIGA_SYM_MODE_HMAC_SHA256, OID_PIN_HMAC, digest,
+                           sizeof(digest), hmac_buffer, sizeof(hmac_buffer),
+                           &size);
+  if (res != OPTIGA_SUCCESS) {
+    memzero(digest, sizeof(digest));
+    uint8_t error_code = 0;
+    optiga_get_error_code(&error_code);
+    if (error_code + OPTIGA_COMMAND_ERROR_OFFSET ==
+        OPTIGA_ERR_ACCESS_COND_NOT_SAT) {
+      return OPTIGA_ERR_COUNTER_EXCEEDED;
+    } else {
+      return error_code + OPTIGA_COMMAND_ERROR_OFFSET;
+    }
+  }
+
+  // Stretch the PIN with the result.
+  hmac_sha256(stretched_pin, OPTIGA_PIN_SECRET_SIZE, hmac_buffer, size,
+              stretched_pin);
+  memzero(hmac_buffer, sizeof(hmac_buffer));
+
+  // Process the stretched PIN using a one-way function before sending it to the
+  // Optiga.
+  hmac_sha256(stretched_pin, OPTIGA_PIN_SECRET_SIZE, NULL, 0, digest);
+
+  // Authorise using OID_STRETCHED_PIN so that we can read from OID_PIN_SECRET
+  // and reset OID_PIN_HMAC_CTR.
+  res = optiga_set_auto_state(OPTIGA_OID_SESSION_CTX, OID_STRETCHED_PIN, digest,
+                              sizeof(digest));
+  memzero(digest, sizeof(digest));
+  if (res == OPTIGA_ERR_CMD) {
+    uint8_t error_code = 0;
+    optiga_get_error_code(&error_code);
+    return error_code + OPTIGA_COMMAND_ERROR_OFFSET;
+  }
+
+  ui_progress(200);
+
+  // Reset the counter which limits the use of OID_PIN_HMAC.
+  res = optiga_set_data_object(OID_PIN_HMAC_CTR, false, COUNTER_RESET,
+                               sizeof(COUNTER_RESET));
+  if (res != OPTIGA_SUCCESS) {
+    return res;
+  }
+
+  // Read the counter-protected PIN secret from OID_PIN_SECRET.
+  uint8_t pin_secret[OPTIGA_PIN_SECRET_SIZE] = {0};
+  res = optiga_get_data_object(OID_PIN_SECRET, false, pin_secret,
+                               OPTIGA_PIN_SECRET_SIZE, &size);
+  if (res != OPTIGA_SUCCESS) {
+    optiga_clear_auto_state(OID_STRETCHED_PIN);
```

Comment:
I would call `memzero(pin_secret)` here.

Reply:
if `res != OPTIGA_SUCCESS` then `pin_secret` is not set. The `optiga_` functions modify the destination buffer only if the command succeeds.